docker/Dockerfile

```diff
@@ -116,39 +120,136 @@ COPY --from=openssl_image /usr/local/lib /usr/local/lib
+    make
+
+# Copy openssl build artifacts from openssl_image
+COPY --from=openssl_image /usr/local/ssl /usr/local/ssl
+COPY --from=openssl_image /usr/local/bin /usr/local/bin
+COPY --from=openssl_image /usr/local/include /usr/local/include
+COPY --from=openssl_image /usr/local/lib /usr/local/lib
+
+RUN ldconfig \
+ && ln -s /etc/ssl/certs/* /usr/local/ssl/certs/
+
+COPY VERSION /project/avalon/
+COPY ./bin /project/avalon/bin
+
+ENV TCF_HOME=/project/avalon
+
+COPY ./common/python /project/avalon/common/python
+
 WORKDIR /project/avalon/common/python
 
 RUN echo "Building Avalon Common Python\n" \
-  && make
+ && make
+
+
+
+# -------------=== common/crypto_utils build ===-------------
+
+#Build common/cpp intermediate docker image
+FROM python_image as common_crypto_utils_image
+
+RUN apt-get update \
+ && apt-get install -y -q \
+    swig \
+    make
+
+
+# Copy openssl build artifacts from openssl_image
+COPY --from=openssl_image /usr/local/ssl /usr/local/ssl
+COPY --from=openssl_image /usr/local/bin /usr/local/bin
+COPY --from=openssl_image /usr/local/include /usr/local/include
+COPY --from=openssl_image /usr/local/lib /usr/local/lib
+
+RUN ldconfig \
+ && ln -s /etc/ssl/certs/* /usr/local/ssl/certs/
+
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-base64.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-common.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-crypto.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-parson.a /project/avalon/common/cpp/build/
+COPY --from=common_cpp_image /project/avalon/common/cpp/build/libuavalon-verify-ias-report.a /project/avalon/common/cpp/build/
+COPY VERSION /project/avalon/
+COPY ./bin /project/avalon/bin
+COPY ./common/cpp /project/avalon/common/cpp
+
+ENV TCF_HOME=/project/avalon
+
+COPY ./common/crypto_utils /project/avalon/common/crypto_utils
 
 WORKDIR /project/avalon/common/crypto_utils
 
 RUN echo "Building Avalon Common Crypto Python\n" \
-  && make
+ && make
+
+
+
+# -------------=== Avalon SDK build ===-------------
+
+#Build Avalon SDK intermediate docker image
+FROM python_image as avalon_sdk_image
+
+RUN apt-get update \
+ && apt-get install -y -q \
+    make
+
+COPY VERSION /project/avalon/
+COPY ./bin /project/avalon/bin
+COPY ./sdk /project/avalon/sdk
+
+ENV TCF_HOME=/project/avalon
 
 WORKDIR /project/avalon/sdk
 
+#install Avalon SDK  package.
 RUN echo "Building Avalon SDK\n" \
-  && make
+ && make
+
+
 
 # Build Final image and install dependent modules
 FROM base_image as final_image
 
+RUN apt-get update \
+ && apt-get install -y -q \
+    git
+
+# Installing solidity compiler
+RUN add-apt-repository ppa:ethereum/ethereum \
+   && apt-get update \
+   && apt-get install solc
+
+RUN apt-get install -y -q python3-pip \
+ && pip3 install --upgrade json-rpc web3 py-solc-x nose2 \
```

Comment:
using upgrade may not be good idea as it will always out date the cache and build image every time.